src/main/java/ro/msg/learning/shop/strategy/MostAbundantStrategy.java

```diff
@@ -0,0 +1,51 @@
+package ro.msg.learning.shop.strategy;
+
+import lombok.AllArgsConstructor;
+import ro.msg.learning.shop.dto.ProductQuantityDto;
+import ro.msg.learning.shop.dto.StockDto;
+import ro.msg.learning.shop.exception.ResourceNotFoundException;
+import ro.msg.learning.shop.model.Stock;
+import ro.msg.learning.shop.repository.ProductRepository;
+import ro.msg.learning.shop.repository.StockRepository;
+import ro.msg.learning.shop.util.StockMapper;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+@AllArgsConstructor
+public class MostAbundantStrategy implements LocationStrategy {
+
+    private final StockRepository stockRepository;
+    private final ProductRepository productRepository;
+    private final StockMapper stockMapper;
+
+    @Override
+    public List<StockDto> findLocation(List<ProductQuantityDto> products) {
+
+        Map<UUID, StockDto> map = new HashMap<>();
+
+        products.forEach(p -> {
+
+            List<Stock> availableStocks = stockRepository.findByProductAndQuantity(
+                    productRepository.findById(p.getProductId())
+                            .orElseThrow(() -> new ResourceNotFoundException("Product not found")), p.getQuantity());
+
+            availableStocks.forEach(s -> {
```

Comment:
You can remove this logic (which I understand that it saves in the map the stock with the maximum value), and you can make the query in such a way that it takes the stocks with the maximum value from the DB directly.